modules/auxiliary/gather/zoomeye_search.rb

```diff
@@ -113,66 +121,148 @@ def dork_search(dork, resource, page)
-      results = dork_search(dork, resource, page) if dork
-      break unless match_records?(results)
+    # Determine page count based on total results
+    if results[first_page]['total'] % 20 == 0
+      tpages = results[first_page]['total'] / 20
+    else
+      tpages = results[first_page]['total'] / 20 + 1
+    end
+    maxpage = tpages if datastore['MAXPAGE'] > tpages
 
-      matches = results['matches']
-      if resource.include?('web')
-        parse_web_records(matches)
+    if facets
+      facets_tbl = Rex::Text::Table.new(
+        'Header' => 'Facets',
+        'Indent' => 1,
+        'Columns' => ['Facet', 'Name', 'Count']
+      )
+      print_status("Total: #{results[first_page]['total']} on #{tpages} " \
+        'pages. Showing facets')
+      facet = results[first_page]['facets']
+      facet.each do |name, list|
+        list.each do |f|
+          facets_tbl << [name.to_s, (f['name']).to_s, (f['count']).to_s]
+        end
+      end
+    else
+      print_status("Total: #{results[first_page]['total']} on #{tpages} " \
+        "pages. Showing: #{maxpage} page(s)")
+      # If search results greater than 20, loop & get all results
+      if results[first_page]['total'] > 20
+        print_status('Collecting data, please wait...')
+        page = 1
+        while page < maxpage
+          page_result = dork_search(resource, dork, page + 1, facets)
+          if page_result['matches'].nil?
+            next
+          end
+          results[page] = page_result
+          page += 1
+        end
+      end
+      tbl1 = Rex::Text::Table.new(
+        'Header' => 'Host search',
+        'Indent' => 1,
+        'Columns' => ['IP:Port', 'City', 'Country', 'Hostname', 'OS', 'Service:Version', 'Info']
+      )
+      tbl2 = Rex::Text::Table.new(
+        'Header' => 'Web search',
+        'Indent' => 1,
+        'Columns' => ['IP', 'Site', 'City', 'Country', 'DB:Version', 'WebApp:Version']
+      )
+      # scroll max pages from ZoomEye
+      results.each do |result|
+        result['matches'].each do |match|
+          city = match['geoinfo']['city']['names']['en']
+          country = match['geoinfo']['country']['names']['en']
+          if resource.include?('host')
+            ip = match['ip']
+            port = match['portinfo']['port']
+            hostname = match['portinfo']['hostname']
+            os = match['portinfo']['os']
+            service = match['portinfo']['app']
+            name = match['portinfo']['name']
+            version = match['portinfo']['version']
+            info = match['portinfo']['extrainfo']
+            if datastore['DATABASE']
+              report_host(host: ip,
+                          name: hostname,
+                          os_name: os,
+                          comments: 'Added from Zoomeye')
+            end
+            if datastore['DATABASE']
+              report_service(host: ip,
+                             port: port,
+                             proto: name,
```

Comment:
This is failing as `name` is not within the included list for some reason, which is causing the stack trace that was shown earlier.

Reply:
See https://github.com/rapid7/metasploit_data_models/blob/31e8b28c95251e6f52bd011c52feda96df9c217f/app/models/mdm/service.rb#L10

And https://github.com/rapid7/metasploit_data_models/blob/31e8b28c95251e6f52bd011c52feda96df9c217f/app/models/mdm/service.rb#L233-L236

Valid values for `proto` are `udp` or `tcp`

Reply:
Leaving this open as I still think there is a legitimate case here for adding in `proto` in some way. We should be able to get this information so the question is why are we not getting it and how can we update this properly? If we are supplying `port` we should also be able to tell what protocol it is, otherwise there is no way to use the information later on as we won't know if it was a TCP port or a UDP port.

Reply:
Also the `proto` is actually required as noted at https://github.com/rapid7/metasploit-framework/blob/16923f4962fc67f9779207d11f413b4ec011c9d3/lib/msf/core/db_manager/service.rb#L36-L51

Reply:
Fixed in https://github.com/rapid7/metasploit-framework/pull/15278/commits/5b30e54b102dcb3b973056b91eea46d751a6d060